libc/src/__support/str_to_float.h

```diff
@@ -1208,6 +1210,15 @@ template <class T> LIBC_INLINE StrToNumResult<T> strtonan(const char *arg) {
   using FPBits = typename fputil::FPBits<T>;
   using StorageType = typename FPBits::StorageType;
 
+#ifndef LIBC_HAS_SANITIZER
+  if (LIBC_UNLIKELY(arg == nullptr)) {
+    // Use volatile to prevent undefined behavior of dereferencing nullptr.
```

Comment:
given that this problem is only likely to occur in the tests, does it make more sense to change the tests instead of adding this extra-explicit check here?

Reply:
No, the main purpose is not for the test, but hardening user code when sanitizers are not in used.  This is especially important when this header function is included and built from source, as the compiler can detect undefined behavior and optimize the code away completely.  The explicit check is for performance purpose, since you will not want everything to always be round-tripped through the volatile variable.

Reply:
From what I understand, the only time this will matter is when all of the following conditions are met:
a) User code is built alongside libc code as source
b) The code is built without sanitizers
c) The code is built with optimizations
d) The code passes a nullptr to the `nan` function in a constant way.

In that case, the user code is incorrect (it should not pass nullptr to `nan`) and the compiler may do unexpected things. I agree that this check would help find the bug in this specific case, but it will have a penalty for every other case, since this check can't be optimized out due to the `volatile`. Additionally, if the user code does have this undefined behavior, then they can find it by changing any of the conditions (e.g. building with sanitizers) without us needing this check all the time.

My main point is: If the user wants hardening to find this sort of undefined behavior, they have access to it. If they didn't choose to use hardening then they may not want it.

Reply:
It's triggered even without a & d.  This got my attention when the nan tests didn't crash and trigger SIGSEGV on some riscv64 configs.

About your worry regarding the performance, if the inputs (runtime or compile-time) have nullptr, our original intention is to crash with SIGSEGV anyway, and the added `nullptr` check and volatile load is negligible compared to crashing part.
If the inputs do not have nullptr, the only extra cost is the nullptr check `arg == nullptr`, which always returns false under the condition / assumption, cost at most 1 CPU clock, and will be hidden away completely by the CPU's branch predictors.

So for this check `arg == nullptr` to incurs observable performance penalty, you'll need to call `strtonan` for a large array with inputs alternating between very short string and `nullptr` in order to beat the CPU's branch predictor.  But in that case, we already crash at the first `nullptr` encountered.

Reply:
That's a fair point. In that case we should create a generic way to do nullptr checks so it can be used in other places, and also guard it behind a hardening macro so it can be turned off on space-sensitive systems.

Reply:
I added a config option to enable/disable null checks.  At the current form it's not straightforward to refactor it to a generic macro.  I'll leave adding a generic macro for later, after seeing how to do null checks in other functions.

Reply:
I'd suggest replacing the `volatile` stuff with just `__builtin_trap()`. It makes this more generic and makes it less dependent on optimization behavior. If you really want it to segfault, then you could do this: 
```C++
volatile char *crashing = reinterpret_cast<char*>(arg); //might also need a const_cast to be properly generic
char a = *crashing;
__builtin_trap(); //avoid the return while also making it clear to the compiler that this is a terminal path
```

Reply:
I refactor it to a generic macro in `__support/macros/null_check.h`